src/runtime/hexagon_dma.cpp

```diff
@@ -121,6 +127,150 @@ static void desc_pool_free (void* user_context) {
     }
 }
 
+static inline void copy_from_cache(uint8_t* src, uint8_t* dest, int roi_height,
+           int roi_stride, int dest_stride, int pixelsize, int linesize) {
+     
+    for (int y = 0; y < roi_height; y++) {
+        int sy = y * roi_stride * pixelsize;
+        int dy = y * dest_stride * pixelsize;
+        memcpy(&dest[dy], &src[sy], linesize);
+    }
+} 
+
+static int halide_hexagon_dma_wrapper (void *user_context, struct halide_buffer_t *src,
+                                       struct halide_buffer_t *dst) {
+
+    dma_device_handle *dev = (dma_device_handle *)src->device;
+
+    debug(user_context)
+        << "Hexagon dev handle: buffer: " << dev->buffer << " dev offset (" << dev->offset_x << ", " << dev->offset_y << ") frame_width: " << dev->frame_width
+        << " frame_height: " << dev->frame_height << " frame_stride: " << dev->frame_stride << "\n";
+
+    int roi_stride = 0;
+    int roi_width = 0;
+    int roi_height = 0;
+
+    // Changing the Format to Chroma or LUMA based on dimension    
+    t_eDmaFmt currentFmt = dev->fmt;
+    if (dst->dimensions == 3) {
```

Comment:
We've got to find another way to give the DMA engine the information it needs. I think the basic problem is that we're trying to do something here that Halide just doesn't understand: you have a single input buffer with a Y plane (W x H) and a UV plane (W/2 x H/2 x 2), and you're trying to read it in a Halide program as two separate buffers. This necessary because Halide doesn't understand buffers with channels of different dimensions, but the approach you are taking is also highly suspect.

I really, really wish we could focus on plain-old-data rather than formats with "built-in implications" like YUV 4:2:0.

Basically, the problem you've undertaken here really is not a DMA engine problem, it's a YUV 4:2:0 problem. I don't have an easy answer for this.

Reply:
Agree that we have to find a different way to address this but currently this is the best solution. Once there is a Halide solution for YUV 420 we will adopt accordingly.

We are using W * H * 2 as an interim solution so that we can flatten the Halide Buffer to allow the DMA to function properly using one buffer.